lib/posix-process: Add vfork & execve

michpappas opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [N/A]
 - Platform(s): [N/A]
 - Application(s): [N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->
* CONFIG_LIBPOSIX_PROCESS_EXECVE: Enables the `execve()` syscall
* CONFIG_LIBPOSIX_PROCESS_VFORK: Enables the `vfork()` syscall
* CONFIG_LIBUKBINFMT: Enables the `ukbinfmt` library

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

This PR adds support for `vfork()` / `execve()` to libposix-process. This provides the foundation for running multiprocess applications on Unikraft.

`vfork()` creates a child process that shares the same memory with the calling thread, including the stack. Upon `vfork()` the parent is suspended until the child calls `execve()` or `_exit()` to prevent the parent's state from being corrupted. Due to above, the only permissible actions taken by the child are:
* Update a variable of type `pid_t` to store its return value.
* Calling any function other than one of the functions of the `exec(2)` family, or `_exit(2)`.

Any action besides the above results into undefined behavior.

For a more elaborate description on `vfork()` see [vfork(2)](https://man7.org/linux/man-pages/man2/vfork.2.html).

`execve()` implements the standard behavior described in [execve(2)](https://man7.org/linux/man-pages/man2/execve.2.html). To provide flexibility on the supported binary formats the implementation relies on `libukbinfmt`, which is also introducedin this series. Similarly to linux, this provides a minimal framework for registering loaders that handle different binary formats, introducing the ability of loading interpreted scripts. The implementation may be extended with support for linux's [`binfmt_misc`](https://docs.kernel.org/admin-guide/binfmt-misc.html) fs, should such a use-case arise. The implementation of the ELF `ukbinfmt` loader is introduced in `app-elfloader`.

GitHub-Depends-On: #1316
GitHub-Depends-On: #1322
GitHub-Depends-On: #1346
GitHub-Depends-On: https://github.com/unikraft/app-elfloader/pull/79

michpappas committed:
lib/posix-process: Add posix_thread state

The state provides information on whether a posix_thread is
running, blocked, or exited.

Notice that posix_thread_state is only updated by operations
at the posix_process / posix_thread level and may not
always be in sync with the state of the underlying uk_thread.
This specifically applies to the POSIX_THREAD_RUNNING state,
which may not be accurate e.g. if the underlying uk_thread
blocks at the scheduler due to a lock.

On the other hand, the variants of POSIX_STATE_BLOCKED always
reflect the state of a posix_thread, as it is certain that the
underlying uk_thread will also be blocked from the scheduler.

Given the above, a check against POSIX_STATE_RUNNING should only be
used to check if the state of a posix-thread is not terminated or
blocked.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/vfscore: Handle CLONE_VM

vfork() sets the CLONE_VM and CLONE_VFORK flags. This triggers an
error in the clone handlers of vfscore as CLONE_FS is not set. Update
the handlers to additionally check against CLONE_VM, as that also
implies that the parent and child share filesystem state.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/posix-process: Migrate tid2pprocess / tid2pthread to process.h

Migrate the definitions of tid2pthread() and tid2pprocess() to
the private process.h to make them available to the rest of the
library. This requires to additionally migrate the definitions of
struct posix_process() and struct posix_thread().

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/nolibc: Add definition of ARG_MAX

Add definition of ARG_MAX to limits.h. POSIX defines ARG_MAX as the
number of bytes available for the combined arguments and env vars
of a new process. Whether that additionally includes NULL terminator,
pointers, or alignment bytes is IMPLEMENTATION DEFINED.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib: Introduce libukbinfmt

libukbinfmt provides a minimal framework to register handlers of
executable files. Typical examples include binary executables like
ELF objects, or interpreted files like *nix scripts that use the
sha-bang sequence to specify an interpreter.

This commit only implements the functionality required to register
and execute loaders within the kernel's scope. Additional
functionality incl. application support via Linux's `binfmt_misc`
API shall be added as a future extension.

Checkpatch-Ignore: COMPLEX_MACRO
Checkpatch-Ignore: MACRO_ARG_REUSE

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/posix-process: Add execve()

Add implementation for execve(). For more info see execve(2).

Requires a binfmt ELF loader. The default loader is provided by
app-bincompat.

Checkpatch-Ignore: AVOID_EXTERNS

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/posix-process: Add vfork()

The vfork() syscall is equivalent to calling clone() with the flags
parameter set to CLONE_VM | CLONE_VFORK | SIGCHLD. Update clone() to
support CLONE_VFORK and CLONE_VM. Implement vfork() as a wrapper of
clone().

For more info see vfork(2).

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/posix-process: Add posix_thread parent

Add a field to posix_thread to keep track of its parent. This is
populated during the creation of a posix_thread, and it is used
for deriving the parent's state in execve() / exit().

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

mogasergiu commented on lib/Makefile.uk:
```diff
@@ -70,3 +70,4 @@ $(eval $(call import_lib,$(CONFIG_UK_BASE)/lib/ukrust))
 $(eval $(call import_lib,$(CONFIG_UK_BASE)/lib/ukreloc))
 $(eval $(call import_lib,$(CONFIG_UK_BASE)/lib/ukofw))
 $(eval $(call import_lib,$(CONFIG_UK_BASE)/lib/ukallocstack))
+$(eval $(call import_lib,$(CONFIG_UK_BASE)/lib/ukbinfmt))
```
> lib: Introduce libukbinfmt

s/sha-bang/shebang in the commit msg

michpappas replied:
"sha-bang" is also valid, see https://en.wikipedia.org/wiki/Shebang_(Unix)

mogasergiu replied:
Ok fine... but I think you can agree with me that we all would much rather have a "she" than a "sha" :innocent: :stuck_out_tongue:

mogasergiu commented on lib/ukbinfmt/include/uk/binfmt.h:
```diff
@@ -0,0 +1,102 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
```
2024

mogasergiu commented on lib/posix-process/process.h:
```diff
@@ -58,6 +67,7 @@ struct posix_thread {
 	struct uk_list_head thread_list_entry;
 	struct uk_thread *thread;
 	struct uk_alloc *_a;
+	enum posix_thread_state state;
```
I would also add a comment here that states again the potential state difference between uk_thread and posix_thread

mogasergiu commented on lib/posix-process/arch/arm64/execve.c:
```diff
@@ -0,0 +1,20 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
```
2024

mogasergiu commented on lib/posix-process/arch/x86_64/execve.c:
```diff
@@ -0,0 +1,29 @@
+
+	/* Prepare for iret */
+	execenv->regs.eflags = X86_EFLAGS_IF; /* enable IRQs on ctx switch */
+	execenv->regs.cs = 8;
```
`GDT_DESC_OFFSET(GDT_DESC_CODE)` and `GDT_DESC_OFFSET(GDT_DESC_DATA)`. yes this introduces deps to plat common :zany_face:

michpappas replied:
I'm not sure we want to introduce such a dependency though. I'll add a `FIXME`.

mogasergiu replied:
~~Can you also add a comment like a TODO or smth so that someone in the future may do this replacement once we get rid of this `plat/common` stuff?~~
Nvm, yes this is what you meant, I somehow skipped the last sentence :man_facepalming:

mogasergiu commented on lib/posix-process/arch/x86_64/execve.c:
```diff
@@ -0,0 +1,29 @@
+#include <uk/arch/ctx.h>
+#include <uk/essentials.h>
+
+void execve_arch_execenv_init(struct ukarch_execenv *execenv,
```
Why no asserts! :angry:

mogasergiu commented on lib/posix-process/arch/arm64/execve.c:
```diff
@@ -0,0 +1,20 @@
+#include <uk/arch/ctx.h>
+#include <uk/essentials.h>
+
+void execve_arch_execenv_init(struct ukarch_execenv *execenv,
```
Ditto :relieved:

mogasergiu commented on lib/posix-process/execve.c:
```diff
@@ -0,0 +1,194 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
```
2024 :clock10:

mogasergiu commented on lib/posix-process/arch/arm64/execve.c:
```diff
@@ -0,0 +1,20 @@
+void execve_arch_execenv_init(struct ukarch_execenv *execenv,
+			      __uptr ip, __uptr sp)
+{
+	execenv->regs.lr = ip;
```
I think you may want to use `elr_el1` here instead.

michpappas replied:
Oh yes

mogasergiu commented on lib/posix-process/execve.c:
```diff
@@ -0,0 +1,194 @@
+	this_thread = uk_thread_current();
+	UK_ASSERT(this_thread);
+
+	pthread = tid2pthread(ukthread2tid(this_thread)); /* FIXME */
```
Should we assert `pthread` too? :thinking: Or we assume it's fine based on the above assertion?

michpappas replied:
I can add an assertion

mogasergiu commented on lib/posix-process/arch/arm64/execve.c:
```diff
@@ -0,0 +1,20 @@
+{
+	execenv->regs.lr = ip;
+	execenv->regs.sp = sp;
+
```
I think we should also set `spsr_el1` as I believe it to have an unsafe value coming from a `malloc`d stack. I would reuse the one of the parent and, additionally ` &= ~PSR_I` so that IRQs are enblaed. Ditto for x86, let's reuse the parents flags but with IRQs forcefully enabled (although I think in both cases it should be safe to assume they are already enabled IMO).

michpappas replied:
Actually, what is the reasoning for updating `SPSR_EL1` and `ESR_EL1` in `ukarch_execenv_load`?

mogasergiu replied:
Following an offline discussion, we decided that we should for now keep these semantics for both `ESR_EL1` and `SPSR_EL1`.

So, initially, the ARM code shared the entries for both IRQ's and regular traps. Since the former would not require `ESR_EL1` and potentially creating confusion we would normally restore `ESR_EL1` to make sure that IRQ code does not somehow think it is in a trap because `ESR_EL1` has a value that a trap would normally set. But I will soon submit a small patch that separates the two entries, thus restoring `ESR_EL1` is no longer required.

On the other hand, we still have to keep `SPSR_EL1` updated as that restores the `PSTATE`. For example, consider a page fault that may happen during a syscall. This is essentially a nested trap so `SPSR_EL1` must remain consistent when returning from the page fault as that could mean confusing the previous trap. Same thing if we nest page faults, which we should soon have a use case for.

So, for now, it is not clear if it is really worth dropping `ESR_EL1`, as we cannot see the use of restoring it anymore but, at the same time, it ensures a little bit of consistency on return from exceptions, although we are not sure if it is a useful consistency to have either, it will probably save us in the future. From what I have seen, neither linux not freebsd restore it on exception return but, I do not have knowledge of how they deal with nested pagefaults, or if they do consider them at all.

mogasergiu commented on lib/posix-process/arch/x86_64/execve.c:
```diff
@@ -0,0 +1,29 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
```
:two: :zero: :two: :four:

mogasergiu commented on lib/posix-process/clone.c:
```diff
@@ -310,19 +310,19 @@ static void _clone_child_gc(struct uk_thread *t)
  */
-static int _clone(struct clone_args *cl_args, size_t cl_args_len,
-		  struct ukarch_execenv *execenv)
+int uk_clone(struct clone_args *cl_args, size_t cl_args_len,
```
`__sz`?

michpappas replied:
Had I written this from scratch yes, but since the rest of this file uses `size_t` I chose to stick to the existing convention.

mogasergiu commented on lib/posix-process/clone.c:
```diff
@@ -388,6 +388,25 @@ static int _clone(struct clone_args *cl_args, size_t cl_args_len,
+	if ((flags & CLONE_VM)) {
+		if (!cl_args->stack && !cl_args->stack_size) {
+			uk_pr_debug("Using parent's sp @ 0x%lx\n", execenv->regs.rsp);
+			cl_args->stack = execenv->regs.rsp; /* FIXME */
```
I assume FIXME cus x86 specific, right?

michpappas replied:
Correct. I now included your patch with `ukarch_regs_get_sp()` to resolve this.

mogasergiu commented on lib/posix-process/clone.c:
```diff
@@ -476,10 +520,21 @@ static int _clone(struct clone_args *cl_args, size_t cl_args_len,
+	 * or exit(). Yield to schedule the child.
+	 */
+	if (flags & CLONE_VFORK) {
+		struct posix_thread *pthread = tid2pthread(ukthread2tid(t));
```
Subscope declaration? So unlike you :disappointed:

mogasergiu reviewed (changes requested):
Hi there :wave: :man_health_worker: . Had a brief go at this. I will have a deeper look once we get the dependency PRs in.

mogasergiu commented on arch/arm/arm64/include/uk/asm/lcpu.h:
```diff
@@ -223,6 +223,11 @@ struct __regs {
 
 UK_CTASSERT(sizeof(struct __regs) == __REGS_SIZEOF);
 
+static inline __uptr ukarch_regs_get_sp(struct __regs *r)
```
Don't forget to add your sob + cab.

mogasergiu commented on lib/ukbinfmt/binfmt.c:
```diff
@@ -0,0 +1,100 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
```
2024

mogasergiu commented on lib/ukbinfmt/binfmt.c:
```diff
@@ -0,0 +1,100 @@
+{
+	UK_ASSERT(loader);
+	UK_ASSERT(loader->name);
+	UK_ASSERT(loader->ops.load);
```
Is it mandatory that both ops are implemented and that it has a name? I assume this is because `uk_binfmt_load` does not check for such cases?

michpappas replied:
Yes, they are all mandatory. The name is useful for messages. For ops, `unload()` is necessary for transient payloads.

mogasergiu commented on lib/ukbinfmt/binfmt.c:
```diff
@@ -0,0 +1,100 @@
+		UK_ASSERT(ldr->ops.load);
+		rc = ldr->ops.load(args);
+		if (unlikely(rc < 0))
+			return rc;
```
Perhaps an err msg?

mogasergiu commented on lib/ukbinfmt/binfmt.c:
```diff
@@ -0,0 +1,100 @@
+
+	uk_binfmt_loader_foreach(ldr, UK_BINFMT_LOADER_TYPE_EXEC) {
+		rc = ldr->ops.load(args);
+		if (unlikely(rc < 0))
```
ditto

mogasergiu commented on lib/ukbinfmt/binfmt.c:
```diff
@@ -0,0 +1,100 @@
+
+	uk_pr_debug("Unload handled by %s\n", args->loader->name);
+	rc = args->loader->ops.unload(args);
+	if (unlikely(rc < 0))
```
ditto

mogasergiu commented on lib/ukbinfmt/include/uk/binfmt.h:
```diff
@@ -0,0 +1,102 @@
+
+struct uk_binfmt_loader_args {
+	const char *pathname; /* fully qualified */
+	char *progname;
```
Would it be worth it documenting each field? Like a 1-line comment, since they are public interfaces.

mogasergiu commented on lib/ukbinfmt/include/uk/binfmt.h:
```diff
@@ -0,0 +1,102 @@
+
+struct uk_binfmt_loader {
+	const char *name;
+	enum uk_binfmt_type type;
```
Ditto

mogasergiu commented on lib/posix-process/Makefile.uk:
```diff
@@ -24,7 +27,7 @@ LIBPOSIX_PROCESS_ASINCLUDES-y += -I$(UK_PLAT_COMMON_BASE)/include
 
 UK_PROVIDED_SYSCALLS-$(CONFIG_LIBPOSIX_PROCESS_CLONE) += clone-5e
-UK_PROVIDED_SYSCALLS-$(CONFIG_LIBPOSIX_PROCESS) += execve-3
+UK_PROVIDED_SYSCALLS-$(CONFIG_LIBPOSIX_PROCESS_EXECVE) += execve-3
```
`execve-3e`? :thinking:

michpappas replied:
I wonder why this worked btw. Unless that `e` was eaten by a confilct during a rebase 🤔

mogasergiu commented on lib/posix-process/arch/arm64/execve.c:
```diff
@@ -0,0 +1,34 @@
+	UK_ASSERT(execenv);
+	UK_ASSERT(ip);
+	UK_ASSERT(sp);
+
```
Perhaps an assertion for sp alignment?

mogasergiu commented on lib/posix-process/arch/arm64/execve.c:
```diff
@@ -0,0 +1,34 @@
+	/* Copy ESR to make sure we restore a sane value */
+	execenv_new->regs.esr_el1 = execenv->regs.esr_el1;
+
+	/* Also copy the current sysregs to avoid ending up with undefined
```
s/sysregs/sysctx

mogasergiu commented on lib/posix-process/arch/x86_64/execve.c:
```diff
@@ -0,0 +1,37 @@
+	execenv_new->regs.rip = ip;
+	execenv_new->regs.rsp = sp;
+
+	/* Prepare for iret
```
Nitpick: let's be specific: `iretq`. As there is also `iret` and `iretd` whose semantics can get very different. Just to avoid confusion of future readers.

mogasergiu commented on lib/posix-process/execve.c:
```diff
@@ -0,0 +1,197 @@
+	 * from init without having called vfork().
+	 */
+	if (!pthread_parent) {
+		UK_ASSERT(pthread->process->pid == 0); /* FIXME */
```
What is this FIXME about?

michpappas replied:
IIRC it's because the posix-process API does not provide accessors so I had to resort to that double indirection. I'll add an assertion to make sure `pthread->process` is not NULL.

mogasergiu commented on lib/posix-process/clone.c:
```diff
@@ -388,6 +389,27 @@ static int _clone(struct clone_args *cl_args, size_t cl_args_len,
+	 * parent's in clone_setup_arch_sysregs(). We copy the registers
+	 * later below.
+	 */
+	if ((flags & CLONE_VM)) {
```
Why double paranthesis?

mogasergiu commented on lib/posix-process/clone.c:
```diff
@@ -388,6 +389,27 @@ static int _clone(struct clone_args *cl_args, size_t cl_args_len,
+	 */
+	if ((flags & CLONE_VM)) {
+		if (!cl_args->stack && !cl_args->stack_size) {
+			cl_args->stack = ukarch_regs_get_sp(&execenv->regs);
```
This is probably just my hunch, but it feels uncomfortable modifying the passed argument here (ditto for tls). Can we instead do this assignment on a locally defined variable instead? Or is there a reason behind this.

michpappas replied:
Yes, I went through that path as well. The problem with using local variables is that it would require making every subsequent access to `cl_args->stack` and `cl_args->tls` conditional to:
```c
if ((flags & CLONE_VM) && (!cl_args->stack))
  /* use local var */
else
 /* use cl_args->stack */

if ((flags & CLONE_VM) && (!cl_args->tls))
  /* use local var */
else
 /* use cl_args->tls */
```
Because these variables are used in several places in `clone()`, that would result in a lot of code duplication which I wanted to avoid as the implementation is complex enough already.

mogasergiu replied:
ok but this still feels awkward and I am thinking of the case where someone may pass something to clone and wake up to clonetab having a different value somehow. Maybe at least there should be a warning/note comment somewhere?

mogasergiu commented on lib/posix-process/clone.c:
```diff
@@ -446,16 +468,46 @@ static int _clone(struct clone_args *cl_args, size_t cl_args_len,
+		 * although a Unikraft TLS was allocated.
+		 */
+		if ((flags & CLONE_SETTLS) || (flags & CLONE_VM))
+			child->tlsp = cl_args->tls;
```
And here we reuse that cl_args argument.

mogasergiu commented on lib/posix-process/clone.c:
```diff
@@ -446,16 +468,46 @@ static int _clone(struct clone_args *cl_args, size_t cl_args_len,
+	}
+
 	/* Call clone handler table but treat CLONE_SETTLS as handled */
 	ret = _uk_posix_clonetab_init(cl_args, cl_args_len,
```
Same here. Does clonetab rely on being preceeded by such operations that would adjust the original user `cl_args`?

mogasergiu commented on lib/posix-process/clone.c:
```diff
@@ -476,10 +528,21 @@ static int _clone(struct clone_args *cl_args, size_t cl_args_len,
 	/* Assign the child to the scheduler */
 	uk_sched_thread_add(s, child);
 
+	/* CLONE_VFORK: Block the parent until the child calls execve()
```
I am not sure this is related to this PR but I have just noticed that `ukthread2tid` is called before `uk_sched_thread_add`. It feels awkward. Do  you happen to know why this order?

michpappas replied:
Maybe I don't fully understand the question, but I don't see a problem. `tid` is assigned by `libposixprocess` at thread creation (https://github.com/unikraft/unikraft/blob/staging/lib/posix-process/process.c#L462) and it's not related to whether the thread is added to the scheduler or not.

mogasergiu replied:
Sorry I got confused by that `ret`. I thought that we are calling a function that returns a value and right afterwards we call a different function without even checking if that `ret` is valid or an error.... but in reality it's just a misnamed local variable... probably `tid` would have been more appropriate instead of `ret`. Anyway it's not realted to this PR I guess...

mogasergiu commented on lib/posix-process/vfork.c:
```diff
@@ -0,0 +1,36 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
```
2024

mogasergiu commented on lib/posix-process/vfork.c:
```diff
@@ -0,0 +1,36 @@
+		return child_tid;
+	}
+
+	return tid2pthread(child_tid)->process->pid;
```
I feel like we should have maybe some assertions or checks here?

michpappas replied:
I'll add an assertion for the child process

mogasergiu commented on lib/posix-process/process.h:
```diff
@@ -41,6 +41,21 @@
 #define TIDMAP_SIZE (CONFIG_LIBPOSIX_PROCESS_MAX_PID + 1)
 
+/* Notice: The RUNNING state is not necessarily in sync with the state
+ * of the undelrying uk_thread (may be blocked by the scheduler).
```
s/undelrying/underlying/
Btw, you also used both `uk_thread` and `uk-thread` in this paragraph.

mogasergiu commented on lib/posix-process/Makefile.uk:
```diff
@@ -9,6 +9,9 @@ CXXINCLUDES-$(CONFIG_LIBPOSIX_PROCESS) += $(LIBPOSIX_PROCESS_COMMON_INCLUDES-y)
 LIBPOSIX_PROCESS_CFLAGS-$(CONFIG_LIBPOSIX_PROCESS_DEBUG) += -DUK_DEBUG
 
+LIBPOSIX_PROCESS_SRCS-$(CONFIG_LIBPOSIX_PROCESS_EXECVE) += $(LIBPOSIX_PROCESS_BASE)/execve.c
+LIBPOSIX_PROCESS_SRCS-$(CONFIG_LIBPOSIX_PROCESS_EXECVE) += $(LIBPOSIX_PROCESS_BASE)/arch/$(CONFIG_UK_ARCH)/execve.c|arch
```
`|arch`? Maybe `|$(CONFIG_UK_ARCH)` or I am missing smth but I do not see `|arch` used anywhere else.

michpappas replied:
That's just namespacing to avoid compile errors, as `execve.o` is already generated for `lib/posix-process/execve.c`. There are no semantics in `arch`, just a way to cause the build system generate `execve.arch.o`.

mogasergiu commented on lib/posix-process/arch/x86_64/execve.c:
```diff
@@ -0,0 +1,38 @@
+	/* Prepare for iretq
+	 * FIXME re-arch: use GDT macros once moved out of plat/common
+	 */
+	execenv_new->regs.eflags = X86_EFLAGS_IF; /* enable IRQs on ctx switch */
```
I don't think this is necessary anymore: <https://github.com/unikraft/unikraft/blob/staging/plat/common/x86/syscall.S#L188>? Either way, not sure about direct assignment, but maybe a `|=` instead.

michpappas replied:
The thing is that we need to initialize it to something as currently the new context contains whatever happens to be in the stack, where it's allocated from. And that's why we also can't do `|=`.

michpappas replied:
What I will do is copy the state from `execenv` instead.

mogasergiu commented on lib/posix-process/arch/x86_64/execve.c:
```diff
@@ -0,0 +1,38 @@
+	execenv_new->regs.ss = 16; /* GDT_DESC_OFFSET(GDT_DESC_DATA) */
+
+	/* Copy current ectx to inerhit platform-initialized regs like mxcsr */
+	ukarch_ectx_sanitize((struct ukarch_ectx *)&execenv_new->ectx);
```
I assume we do not do this on ARM because there we simply copy the Neon regs which are scratch registers as per ABI anyway? If yes we should for sure leave a comment on the ARM implementation as to why we do not do anything ECTX related there.

michpappas replied:
Yes, we don't copy them though. We leave them in an undefined state similarly to the gpregs. I'll add a comment.

mogasergiu commented on lib/posix-process/execve.c:
```diff
@@ -0,0 +1,198 @@
+	/* Assign the default allocator to the loader. This will be used
+	 * to allocate memory for the executable image.
+	 */
+	loader_args.alloc = uk_alloc_get_default();
```
I am wondering... should we maybe use current thread's allocator instead? In practice for now this will be the same thing sure but maybe in the future we will have a use case for separate per-thread allocators.

michpappas replied:
This is intentional. Notice that below I use the thread's allocator for the new stack. The image uses the default allocator similarly to what we do in the `app-elfloader`, so I see it more as a matter of policy. It also avoids the potential issue of the image occupying a significant amount of the thread's dynamic memory pool.

mogasergiu commented on lib/posix-process/arch/x86_64/execve.c:
```diff
@@ -0,0 +1,38 @@
+	UK_ASSERT(sp);
+	UK_ASSERT(IS_ALIGNED(sp, UKARCH_SP_ALIGN));
+
+	execenv_new->regs.rip = ip;
```
Where do you init the other gp regs? Ditto for ARM.

michpappas replied:
I think it's okay to leave gpregs into an undefined state, as we jump into a new execution context.

mogasergiu commented on lib/posix-process/execve.c:
```diff
@@ -0,0 +1,198 @@
+	 * internal cleanup.
+	 */
+	event_data.thread = this_thread;
+	uk_raise_event(POSIX_PROCESS_EXECVE_EVENT, &event_data);
```
This may return an error code.

mogasergiu reviewed (approved):
Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

michpappas commented:
@mogasergiu 
> ok but this still feels awkward and I am thinking of the case where someone may pass something to clone and wake up to clonetab having a different value somehow. Maybe at least there should be a warning/note comment somewhere?

I pushed an update where I shadow elements of `cl_args` into locals.

michpappas commented:
Rebased from `staging`.

michpappas committed:
lib/posix-fdtab: Clear CLOEXEC descriptors upon execve

Add an event handler for POSIX_PROCESS_EXECVE_EVENT to clear
upon execve() file descriptors with CLOEXEC set.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

andreittr commented on lib/ukbinfmt/include/uk/binfmt.h:
```diff
@@ -0,0 +1,106 @@
+
+/** Loader type */
+enum uk_binfmt_type {
+	UK_BINFMT_LOADER_TYPE_NONE,
```
Is there any use for having the enum value `UK_BINFMT_LOADER_TYPE_NONE`? If not, we can consider removing it.

The only use I can think of is ensuring calling code doesn't forget to set loader type when populating a `struct uk_binfmt_loader` via designated initializers. And for this purpose we should explicitly set this enum value to 0, so it gets caught by the assert in `uk_binfmt_register`.

I'm fine with either keeping (w/ explicit value of 0) or ditching, don't feel too strongly either way.

michpappas replied:
Yes exactly, the only reason it's there is to make the check in `uk_binfmt_register()` easy. I think the value does not need to be explicitly initialized to zero, as that's guaranteed by the language.  From [ISO/IEC 9899](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf) Sect 6.7.2.2:
```
The identifiers in an enumerator list are declared as constants that have type int and
may appear wherever such are permitted.109) An enumerator with = defines its
enumeration constant as the value of the constant expression. If the first enumerator has
no =, the value of its enumeration constant is 0.
```
An alternative would had been to set remove `UK_BINFMT_LOADER_TYPE_NONE`, explicitly set the first item to something other than zero, and remove the check altogether. Then validity will be enforced by the enum at compile time.
```
enum uk_binfmt_type {
    UK_BINFMT_LOADER_TYPE_NONE  = 1,
    UK_BINFMT_LOADER_TYPE_EXEC,
};
```
Overall it's not much different than what is currently there, and it can't protect from a caller that tries hard to do something stupid, so I don't have a strong opinion. Happy to go with either (or some alternative proposal, if you have one).

andreittr replied:
I'm aware C guarantees the first enum value to be assigned 0, but I'm not so trusting of future developers remembering not to put any other enum before :smile: . Anyway, it's perfectly fine for now.

I do like the compile-time error idea, that's the best time to catch programming errors! And you don't get away with disabling asserts either :D maybe something for a later change, as it won't change the souce-level API any.

andreittr commented on lib/posix-process/Config.uk:
```diff
@@ -34,7 +34,13 @@ endif
+	bool "execve() system call"
+	select LIBPOSIX_PROCESS_PIDS
+	select LIBUKBINFMT
+	select LIBVFSCORE
```
I think you can remove the hard-select of vfscore here, as no part of `posix-process` actually includes or calls into vfscore directly. From what I see `execve` takes paths and passes them to the binfmt loaders.
(Yes, you do need a VFS to eventually resolve those paths, but IMO that's best left to the binfmt implementation that actually does the VFS calls to `select`.)

michpappas replied:
Yup, thanks for catching that. The VFS requirement is on the `app-elfloader` side. Not sure why it's there, probably a leftover from early development. Will remove.

andreittr commented on lib/posix-process/Makefile.uk:
```diff
@@ -24,7 +27,8 @@ LIBPOSIX_PROCESS_ASINCLUDES-y += -I$(UK_PLAT_COMMON_BASE)/include
 UK_PROVIDED_SYSCALLS-$(CONFIG_LIBPOSIX_PROCESS_CLONE) += clone-5e
-UK_PROVIDED_SYSCALLS-$(CONFIG_LIBPOSIX_PROCESS) += execve-3
+UK_PROVIDED_SYSCALLS-$(CONFIG_LIBPOSIX_PROCESS_EXECVE) += execve-3e
+UK_PROVIDED_SYSCALLS-$(CONFIG_LIBPOSIX_PROCESS_VFORK) += vfork-2e
```
I think this vfork syscall made it into the wrong commit when parceling :smile: 
(Don't bother rewriting history to fix this, it's perfectly fine)

michpappas replied:
Will fix, thanks 👍🏼

andreittr commented on lib/posix-process/execve.c:
```diff
@@ -0,0 +1,202 @@
+	if (!pthread_parent) {
+		UK_ASSERT(pthread->process);
+		UK_ASSERT(pthread->process->pid == 0);
+		UK_ASSERT(pthread_parent->state != POSIX_THREAD_BLOCKED_VFORK);
```
This 3rd assert might trigger UB, as you deref `pthread_parent->state` but you only reach this branch `if (!pthread_parent)` :thinking:

michpappas replied:
Yeah that shouldn't be there, removing.

andreittr reviewed (commented):
Nice work, thanks! I've only some very minor comments, then we can get on with testing and merging.

andreittr reviewed (approved):
Looks good to me, thanks!

Reviewed-by: Andrei Tatar <andrei@unikraft.io>

andreittr commented:
Let's get it merged! :tada: 

Approved-by: Andrei Tatar <andrei@unikraft.io>
